File: cs12m/lab7/Dictionary.java
/*
*Jonathan Molina 
*jomolina
*CS12B
* Lab7 
*
*
*this file implements methods from DictionaryInterface
*
*
*
*
*
*
*
*
* 
*/ 









import java.io.*;
//import java.util.Scanner;
import java.util.*;



class Dictionary implements DictionaryInterface{
       
        

       class Node{
		String key, value;
		Node next, left, right;
		Node(String pairA, String pairB){
			key = pairA;
			value = pairB; 
	
                         left = null;
                         right = null;

	    }
     }
    
        
	public Node head;// makes 
        private int numPairs;
	
       //private methods helper functions
       private Node findKey(Node R, String k){
      
      if(R == null || k.compareTo(R.key) == 0){ 
      return R;
      }
      if( (k.compareTo(R.key)) < 0){ 
      return findKey(R.left, k);
      }
      else{ // strcmp(k, R->key)>0
      return findKey(R.right, k);
      }   							                                                               
      }//closes findkey
      
private  Node findParent(Node N, Node R){
   Node P =null;
   if( N!=R ){
      P = R;
      while( P.left != N && P.right!= N ){
         if(N.key.compareTo(P.key)< 0){
            P = P.left;
         }else{
            P = P.right;
          }
      }
   }
   return P;
}

private Node findLeftmost(Node R){
  Node L = R; 
   if( L !=  null ){
     for( L = R ; L.left != null; L = L.left){
     }
    } 
     return L;
    }  

private String printInOrder( Node R){
  /* if( R != null ){
      //StringBuffer sb = new StringBuffer();
      b += printInOrder(b, R.left);
     b += R.key +" "+ R.value +"\n";
     b += printInOrder( b, R.right);
   }
   return b;*/

if( R!= null ){
      printInOrder( R.left);
      System.out.printf( "%s %s\n", R.key, R.value);
      printInOrder( R.right);
   }


return "";
}
private void  deleteAll(Node N){
   if( N != null ){
      deleteAll(N.left);
      deleteAll(N.right);
      
   }
} 
//Public Methods//
public Dictionary(){
   
 
   head  = null;
   numPairs = 0;
}
public boolean isEmpty(){
return(numPairs == 0);

}
public int size(){
return numPairs;
}
public String lookup(String key){
 
if(findKey(head, key)  == null){
  return null;
}else{
 Node H = findKey(head, key);
  return( H == null? null: H.value);
}
}
public void insert( String key, String value)
	     throws KeyCollisionException{
Node A, N , B;
     if(lookup(key) != null ){
       throw new KeyCollisionException("Cannot insert duplicate keys");
     }
   N = new Node(key, value);
   B = null;
   A = head;
   while( A != null ){
      B = A;
      if(key.compareTo(A.key)< 0){ 
      A = A.left;
      }else{
      A = A.right;
      }
   }
   if( B == null ){
    head = N;
   }else if(key.compareTo(B.key) < 0){
    B.left = N;
   }else{ 
   B.right = N;
   }
   numPairs++;
   
    
  

}
public void delete(String key) throws KeyNotFoundException{
   Node N, P, S;
   N = findKey(head, key);
   if( lookup(key) == null ){
      throw new KeyNotFoundException("Cannot Delete Non-existent key");
   }
   if( N.left== null && N.right==null ){
      if( N == head ){
         head = null;
      }else{
         P = findParent(N, head);
         if( P.right== N ) 
         P.right = null;
         else 
         P.left = null;
        
      }
   }else if( N.right == null ){
      if( N == head ){
         head  = N.left;
         
      }else{
         P = findParent(N, head);
         if( P.right==N ) P.right = N.left;
         else P.left = N.left;
        
      }
   }else if( N.left == null ){
      if( N == head ){
         head = N.right;
        
      }else{
         P = findParent(N, head);
         if( P.right==N ) P.right = N.right;
         else P.left = N.right;
         
      }
   }else{  // N->left!=NULL && N->right!=NULL
      S = findLeftmost(N.right);
      N.key = S.key;
      N.value = S.value;
      P = findParent(S, N);
      if( P.right==S ) P.right = S.right;
      else P.left = S.right;
      
   }
   numPairs--;
}

public void makeEmpty(){	
         head = null;
         numPairs = 0;
}	
public String  toString(){
 Node H = head;
 String s = ""; 
return printInOrder( H );
    
    
   
}  

}

File: Dictionary.c
/*Joanthan Molina
*CS12B
*Lab 5
*jomolina
*
*
*Program looks up words contain in an array
*deletes stuff from array
*adds stuff to list 
*/ 
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


//private types 
typedef struct NodeObj{
     char* key;
     char* value;
     struct NodeObj* next;
}NodeObj;

//Node 
typedef NodeObj* Node;

//newNode()
//contructor of NodeType
Node newNode(char* A, char* B){
Node Al = malloc(sizeof(NodeObj));
assert(Al != NULL);
Al->key = A;
Al->value = B;
Al->next = NULL;
return(Al);
}

//freenode 
//destructs Node
void freeNode(Node *pAl){
   if(pAl != NULL && *pAl != NULL){
   free(*pAl);
   *pAl = NULL;
   }

}
//DictionaryObj array
typedef struct DictionaryObj{
  Node head;
  int  numItems;
}DictionaryObj;
//a reference of a Dictionary obj
typedef DictionaryObj* Dictionary; 
//helper function for Dictionary                              
 Node findKey(Dictionary D, char* key){
     Node Al = D->head;
      for(; Al!= NULL; Al = Al->next){
	    if(strcmp(Al->key, key)==0){

                return Al;
            } 
    }	
    return NULL;        
}
//public functions// 
//constructs Dictionary obj
Dictionary newDictionary(void){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D != NULL);
   D->head = NULL;
   D->numItems = 0;
   return D;
}
//free Dictionary
//destructor for Dictionary
void freeDictionary(Dictionary* pD){
   if(pD != NULL && *pD != NULL){
      if( !isEmpty(*pD)) 
      makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
    
   }
}
//isEmpty()
//returns 1(true) if D is empty, 0 (false) otherwise
//pre: none
int isEmpty(Dictionary D){
  if(D == NULL){
    fprintf(stderr, "Stack Error: Calling isEmpty() on Null Dictionary Reference\n");
    exit(EXIT_FAILURE);
   }
   return(D->numItems==0);
}
//size()
//returns the size of the list
//pre: None
int size(Dictionary D){
return(D->numItems);
}
//looks up word
char* lookup(Dictionary D, char* k){
if(findKey(D, k)==NULL){
    return NULL;
  }else{
    Node Al;
    Al = findKey(D, k);
    return Al->value;
  }
}

//inserts keys and values 
void insert(Dictionary D, char* k, char* v){
Node Al = D->head;
if( D == NULL){
   fprintf(stderr, "Dictionary Error: Calling Insert on Null Dictionary reference\n");
   exit(EXIT_FAILURE);
  }else if(lookup(D,k)!=NULL){
    fprintf(stderr, "Dictionary Error: Inserting Duplicate key\n");
    exit(EXIT_FAILURE);
  }else{
    if(D->head == NULL){
      Al = newNode(k,v);
      D->head = Al;
      D->numItems++;
    }else{
      while(Al != NULL){
        if( Al->next == NULL){
          break;
        }
        Al = Al->next;
      }
      Al->next = newNode(k,v);
      D->numItems++;
    }
  }
}
//deletes keys  
void delete(Dictionary D, char* k){
   Node Al = D->head;
   if( D == NULL ){
      fprintf(stderr, "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }else if( lookup(D, k) == NULL ){
      fprintf(stderr, "Dictionary Error: cannot delete none existing key\n");
      exit(EXIT_FAILURE);
    }else{
    if(D->numItems<=1){;
      freeNode( &Al);
    }else{
      if( Al->key == k){
        Node P = D->head;
        D->head = Al->next;
        D->numItems--;
        freeNode(&P);
      }else{
        while(Al->next->key != k){
          Al = Al->next;
        }
        Node P =  Al->next;
        Al->next = Al->next->next;
        freeNode(&P);
        D->numItems--;
      }
    }
  }
}
//returns everything to an empty state
void makeEmpty(Dictionary D){
 D->numItems = 0;
 freeNode(&D->head);
 }

//prints the values and keys 
void printDictionary(FILE* out, Dictionary D){

Node Al = D->head;
  while( Al != NULL){
    fprintf(out,"%s %s \n",Al->key,Al->value);
    Al = Al->next;
  
}
}











  
